bgpd/bgp_packet.c

```diff
@@ -1005,8 +1005,16 @@ static int bgp_collision_detect(struct peer *new, struct in_addr remote_id)
 					return -1;
 				}
 			else {
-				if (ntohl(peer->local_id.s_addr) ==
-				    ntohl(remote_id.s_addr))
+				/* rfc6286:
+				 * If the BGP Identifiers of the peers
+				 * involved in the connection collision
+				 * are identical, then the connection
+				 * initiated by the BGP speaker with the
+				 * larger AS number is preserved.
+				 */
+				if (ntohl(peer->local_id.s_addr)
+					    == ntohl(remote_id.s_addr)
+				    && peer->local_as < peer->as)
 					flog_err(EC_BGP_ROUTER_ID_SAME, "Peer's router-id %s is the same as ours",
 						 inet_ntoa(remote_id));
```

Comment:
This condition only logs here and no return so even if we have peer->as which larger then we simply not print the log but go on to close our newly created session. As per RFC connection initiated by BGP speaker has to close. That check seems missing.

Reply:
It's closed, look below :)

Reply:
@ton31337 there is no return here. Code currently in master is as below. So with the new check we just seem to log is it not?

				if (ntohl(peer->local_id.s_addr) ==
				    ntohl(remote_id.s_addr))
					flog_err(EC_BGP_ROUTER_ID_SAME, "Peer's router-id %s is the same as ours",
						 inet_ntoa(remote_id));

				/* 3. Otherwise, the local system closes newly
				   created
				   BGP connection (the one associated with the
				   newly
				   received OPEN message), and continues to use
				   the
				   existing one (the one that is already in the
				   OpenConfirm state). */
				if (CHECK_FLAG(peer->sflags,
					       PEER_STATUS_ACCEPT_PEER)) {
					bgp_notify_send(
						peer, BGP_NOTIFY_CEASE,
						BGP_NOTIFY_CEASE_COLLISION_RESOLUTION);
					return 1;
				} else {
					bgp_notify_send(
						new, BGP_NOTIFY_CEASE,
						BGP_NOTIFY_CEASE_COLLISION_RESOLUTION);
					return -1;
				}

Reply:
fixed